ASoC: SOF: ipc4: Add support for core_id in create pipeline message

ujfalusi opened the pull request:
Hi,

the recently merged https://github.com/thesofproject/sof/pull/6637 added support for specifying the target core_id for the pipeline to be created by carving out 4 bits from the reserved section of the message extension.

The core_id can be specified in topology with SOF_TKN_SCHED_CORE token and the kernel will set it to the message.
The core_id will remain 0 (as it is now) when the token is not present and since these bits were ignored, old firmware should with new topology (which is unsupported combination) should not cause regressions, issues either.

replaces PR#4047

Added one new patch to fix a NULL pointer dereference if I force the core_id of the pipelines to 1 which will fail in firmware and the pipeline list allocated, but we have not stored any pipeline in the array, so on stop we are hitting a NULL pointer on the first spipe.

plbossart commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -299,7 +299,7 @@ static int sof_ipc4_trigger_pipelines(struct snd_soc_component *component,
 
 	/* nothing to trigger if the list is empty */
-	if (!pipeline_list->pipelines)
+	if (!pipeline_list->pipelines || !pipeline_list->count)
```
commit message typo: pipleine

ujfalusi replied:
OK.

plbossart commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -675,6 +677,7 @@ static int sof_ipc4_widget_setup_comp_pipeline(struct snd_sof_widget *swidget)
 	pipeline->msg.primary |= SOF_IPC4_MSG_TARGET(SOF_IPC4_FW_GEN_MSG);
 
 	pipeline->msg.extension = pipeline->lp_mode;
+	pipeline->msg.extension |= SOF_IPC4_GLB_PIPE_EXT_CORE_ID(pipeline->core_id);
```
I still don't understand what happens if the pipeline core_id is not aligned with the core_id used for the components of said pipeline, e.g. for DP processing.

ujfalusi replied:
I will update with some protection, I just need to find the correct place.

ujfalusi commented:
Changes since v1:
- typo fixed in the first patch's commit message
- check added to validate the core_id consistency along the path:
 - checking  each connection core_id for pipelines and widgets which will eventually going to find out if any component uses different core_id.

plbossart commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -2106,10 +2111,27 @@ static int sof_ipc4_route_setup(struct snd_sof_dev *sdev, struct snd_sof_route *
+			core_id, src_pipe_widget->widget->name,
+			sink_pipeline->core_id, sink_pipe_widget->widget->name);
+		return -EINVAL;
+	}
```
I don't think this rule is correct. It's perfectly fine to have a route between components that are handled by separate pipelines on separate cores. This would only allow for connected pipelines to be on the same core, and that's a short-term limitation we absolutely don't want to upstream.

plbossart commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -2106,10 +2111,27 @@ static int sof_ipc4_route_setup(struct snd_sof_dev *sdev, struct snd_sof_route *
+			core_id, src_widget->core, src_widget->widget->name,
+			sink_widget->core, sink_widget->widget->name);
+		return -EINVAL;
+	}
```
same on the source side.

ujfalusi replied:
right, so no limitation on the connected pipelines and widgets.
The only limitation is that a widget's core must match with the pipeline's core that it is part of?

ujfalusi replied:
iow, this is not valid:
```c
if (src_pipeline->core_id != src_widget->core) {
...
}

if (sink_pipeline->core_id != sink_widget->core) {
...
}

```

plbossart replied:
> The only limitation is that a widget's core must match with the pipeline's core that it is part of?

Even that part is not clear. When we have DP modules, we will be able to have chained modules on the same pipeline use different cores. So I don't know what the pipeline core would mean, it's an irrelevant notion when DP is used.

ujfalusi replied:
right, so if all combinations are fine then what to check?

ujfalusi replied:
and most importantly, how the DP is going to be used? Is it going to have some mark in tplg?
Without DP we do have limitation, no? We should have the limitation checked now (we don't have DP support afaik) and make it conditional when DP is available?

plbossart replied:
IIRC there is nothing advertized anywhere in kernel-accessible manifests or topology that a module is LL or DP, so it could even be the case that only firmware can check for invalid configurations. @lgirdwood can we have the firmware folks confirm what errors cases we could/need to handled in the driver?

ujfalusi replied:
fwiw, if you send the `CREATE_PIPELINE` message with core_id=1, the firmware will reply with "invalid core id"

plbossart reviewed (changes requested):
I don't think the protection rules are correct.

ujfalusi commented:
Changes since v2:
- Drop the core id verification from `sof_ipc4_route_setup()`
 - the check is going to happen in the firmware and in case of error it is going to be communicated to the kernel which we handle gracefully. This will allow policy and changes in firmware without a need for a synchronized kernel change.

plbossart reviewed (approved):
I would add a comment in the commit message to reflect the direction on error check


Firmware will double-check all information retrieved by topology and report errors if required. This will allow policy and changes in topologies without a need for a synchronized kernel change.

ranj063 commented on sound/soc/sof/ipc4-pcm.c:
```diff
@@ -299,7 +299,7 @@ static int sof_ipc4_trigger_pipelines(struct snd_soc_component *component,
 
 	/* nothing to trigger if the list is empty */
-	if (!pipeline_list->pipelines)
+	if (!pipeline_list->pipelines || !pipeline_list->count)
```
@ujfalusi is this correct? count is only incremented in https://github.com/thesofproject/linux/blob/d39118493a36bb285c10fd763aa3f0954957b46a/sound/soc/sof/ipc4-pcm.c#L65. So it will always be 0 at first no?

ujfalusi replied:
it is incremented in sof_set_up_widgets_in_path():
https://github.com/thesofproject/linux/blob/5694a58bb1e702c05b43134021d382df53529735/sound/soc/sof/sof-audio.c#L465

it is only going to be zero if the sof_widget_setup() earlier fail for the first widget
https://github.com/thesofproject/linux/blob/5694a58bb1e702c05b43134021d382df53529735/sound/soc/sof/sof-audio.c#L445

ranj063 replied:
ahh yes sorry I confused it with the trigget_list->count. This isnt the same

ujfalusi commented:
Changes since v3:
- Commit message updated for the last (core_id support) patch.
